src/sinks/gcp/chronicle_unstructured.rs

```diff
@@ -0,0 +1,521 @@
+        let sink = self.build_sink(client, endpoint, creds)?;
+
+        Ok((sink, healthcheck))
+    }
+
+    fn input(&self) -> Input {
+        Input::log()
+    }
+
+    fn sink_type(&self) -> &'static str {
+        NAME
+    }
+
+    fn acknowledgements(&self) -> Option<&AcknowledgementsConfig> {
+        Some(&self.acknowledgements)
+    }
+}
+
+impl ChronicleUnstructuredConfig {
+    fn build_sink(
+        &self,
+        client: HttpClient,
+        base_url: String,
+        creds: GcpAuthenticator,
+    ) -> crate::Result<VectorSink> {
+        use crate::sinks::util::service::ServiceBuilderExt;
+
+        let request = self.request.unwrap_with(&TowerRequestConfig {
+            rate_limit_num: Some(1000),
+            ..Default::default()
+        });
+
+        let batch_settings = self.batch.into_batcher_settings()?;
+
+        let partitioner = self.key_partitioner()?;
+
+        let svc = ServiceBuilder::new()
+            .settings(request, GcsRetryLogic)
+            .service(ChronicleService::new(client, base_url, creds));
+
+        let request_settings = RequestSettings::new(self)?;
+
+        let sink = GcsSink::new(svc, request_settings, partitioner, batch_settings);
+
+        Ok(VectorSink::from_event_streamsink(sink))
+    }
+
+    fn key_partitioner(&self) -> crate::Result<KeyPartitioner> {
+        Ok(KeyPartitioner::new(self.log_type.clone()))
+    }
+
+    fn endpoint(&self, path: &str) -> Result<String, ChronicleError> {
+        Ok(format!(
+            "{}/{}",
+            match (&self.endpoint, self.region) {
+                (Some(endpoint), None) => endpoint.trim_end_matches('/'),
+                (None, Some(region)) => region.endpoint(),
+                _ => return Err(ChronicleError::RegionOrEndpoint),
+            },
+            path
+        ))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ChronicleRequest {
+    pub body: Bytes,
+    pub finalizers: EventFinalizers,
+    pub metadata: RequestMetadata,
+}
+
+impl Finalizable for ChronicleRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.finalizers)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ChronicleEncoder {
+    customer_id: String,
+    encoder: codecs::Encoder<()>,
+    transformer: codecs::Transformer,
+}
+
+impl Encoder<(String, Vec<Event>)> for ChronicleEncoder {
+    fn encode_input(
+        &self,
+        input: (String, Vec<Event>),
+        writer: &mut dyn io::Write,
+    ) -> io::Result<usize> {
+        let (partition_key, events) = input;
+        let mut encoder = self.encoder.clone();
+        let events = events
+            .into_iter()
+            .filter_map(|mut event| {
+                let timestamp = event
+                    .as_log()
+                    .get(log_schema().timestamp_key())
+                    .and_then(|ts| ts.as_timestamp())
+                    .cloned();
```

Comment:
If we don't clone it `timestamp` is a reference into `event`. This then clashes with the line:

```rust
self.transformer.transform(&mut event);
```

because theoretically `transform` could drop what `timestamp` is pointing to and thus leave it as a reference to nowhere.

There's similar problems no matter where I try to put that line in the function so I don't seem to have a choice but to clone.